docs/architecture/adr-027-deterministic-protobuf-serialization.md

```diff
@@ -0,0 +1,247 @@
+omitted:
+
+```
+$ echo 0a1654686520776f726c64206e65656473206368616e676518e8bebec8bc2e280138024a084e696365206f6e654a095468616e6b20796f75 | xxd -r -p | protoc --decode_raw
+1: "The world needs change"
+3: 1596806111080
+5: 1
+7: 2
+9: "Nice one"
+9: "Thank you"
+```
+
+## Consequences
+
+Having such an encoding available allows us to get deterministic serialization
+for all protobuf documents we need in the context of Cosmos SDK signing.
+
+### Positive
+
+- Well defined rules that can be verified independent of a reference
+  implementation
+- Simple enough to keep the barrier to implement transaction signing low
+- It allows us to continue to use 0 and other empty values in SignDoc, avoiding
+  the need to work around 0 sequences. This does not imply the change from
+  https://github.com/cosmos/cosmos-sdk/pull/6949 should not be merged, but not
+  too important anymore.
+
+### Negative
+
+- When implementing transaction signing, the encoding rules above must be
+  understood and implemented.
+- The need for rule number 3. adds some complexity to implementations.
+
+### Neutral
+
+## References
+
+- <sup>1</sup> _When a message is serialized, there is no guaranteed order for
+  how its known or unknown fields should be written. Serialization order is an
+  implementation detail and the details of any particular implementation may
+  change in the future. Therefore, protocol buffer parsers must be able to parse
+  fields in any order._ from
+  https://developers.google.com/protocol-buffers/docs/encoding#order
+- <sup>2</sup> _Note that for scalar message fields, once a message is parsed
+  there's no way of telling whether a field was explicitly set to the default
+  value (for example whether a boolean was set to false) or just not set at all:
+  you should bear this in mind when defining your message types. For example,
+  don't have a boolean that switches on some behaviour when set to false if you
+  don't want that behaviour to also happen by default._ from
+  https://developers.google.com/protocol-buffers/docs/proto3#default
+- <sup>3</sup> _When a message is parsed, if the encoded message does not
+  contain a particular singular element, the corresponding field in the parsed
+  object is set to the default value for that field._ from
+  https://developers.google.com/protocol-buffers/docs/proto3#default
+- <sup>4</sup> _Also note that if a scalar message field is set to its default,
+  the value will not be serialized on the wire._ from
+  https://developers.google.com/protocol-buffers/docs/proto3#default
+- <sup>5</sup> _For enums, the default value is the first defined enum value,
+  which must be 0._ from
+  https://developers.google.com/protocol-buffers/docs/proto3#default
+- <sup>6</sup> _For message fields, the field is not set. Its exact value is
+  language-dependent._ from
+  https://developers.google.com/protocol-buffers/docs/proto3#default
+- Encoding rules and parts of the reasoning taken from
```

Comment:
I was reading the rendered MD, this should be in italics too?

Reply:
The idea here is that the ones above are quotes and the last one is not. Does this make sense or did I miss something?

Reply:
Ah, got it! You didn't miss anything, I missed that 👍